### PR TITLE
Design review for GRPC implementation

### PR DESCRIPTION
**Pull request by scotte**
Soliciting feedback on this design/approach. Having a separate GRPC object might be overkill, but I was thinking it would make testing easier than if it was built in to the default workload client so a mock GRPC implementation could be used. This might not even make sense since the actual calls use protobuf stubs anyway and I'm not sure what would have to be implemented in the mock to make that work.

Anyway, soliciting feedback on this approach - this pull request is for review only.

### CODE REVIEW

**Review by glaucimar — COMMENTED** [review]
I like this approach if it helps with testing.

File: Pipfile [hunk 1]
```diff
@@ -4,8 +4,8 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-grpcio = "~=1.33.2"
-grpcio-tools = "~=1.33.2"
+grpcio = "~=1.35.0"
+grpcio-tools = "~=1.35.0"
```
[the remark this answers is not in the document]
Comment (scotte) [reply Pipfile:8]: Just updating the dependencies to the latest release version. 1.35 isn't strictly required, but I don't see a reason to use the older one.

File: src/pyspiffe/grpc/default_grpc_client.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from pyspiffe.grpc.grpc_client import GrpcClient
+from grpc import Channel
+from grpc import insecure_channel as grpc_insecure_channel
+
```
[the remark this answers is not in the document]
Comment (scotte) [reply src/pyspiffe/grpc/default_grpc_client.py:4]: Correct, not going to bother with docs at this point.

File: test/grpc/test_default_grpc_client.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from pyspiffe.grpc.default_grpc_client import DefaultGrpcClient
+
+
+def test_instantiate():
+    channel = DefaultGrpcClient.insecure_channel(spiffe_socket='gonzo')
+    channel.close()
```
[the remark this answers is not in the document]
Comment (scotte) [reply test/grpc/test_default_grpc_client.py:6]: It's really just kind of a stub test that no exception gets thrown.

**Review by maxlambrecht — COMMENTED** [review]
Thanks @scotte for this! 
I would like to look deeper into the python/grpc testing capabilities, because maybe we could test the `WorkloadApiClient` by using a stub grpc channel that "connects" to a `FakeWorkloadApi` that responds whatever we need for the tests. If we can do that, we would have in the `WorkloadApiClient` a `workload_pb2_grpc.SpiffeWorkloadAPIStub` and a `grpc_insecure_channel`. It's just opinion and perhaps a little biased by the way I did it in the `java-spiffe`. 😁

File: src/pyspiffe/workloadapi/default_workload_api_client.py [hunk 4]
```diff
@@ -35,6 +44,10 @@ def __init__(self, spiffe_socket: str = None) -> None:
                 'SPIFFE socket argument or environment variable invalid in DefaultWorkloadApiClient.'
             )
 
+        self._channel = grpc_client.insecure_channel(
+            spiffe_socket=self._config.spiffe_endpoint_socket
+        )
+
```
Comment (Andres-GC) [inline src/pyspiffe/workloadapi/default_workload_api_client.py:50]: I'm a little confuse here. 
The way I imagine this is with `DefaultWorkloadApiClient` having a property `grpc_client` that receives the `SpiffeWorkloadAPIStub` already created. So the creation of the channel should be part of the initialization of the `GrpcClient`.

```
class DefaultGrpcClient(GrpcClient):
    def __init__(spiffe_socket: str) -> None:
        channel = grpc_insecure_channel(spiffe_socket)
        return SpiffeWorkloadAPIStub(channel)
```
And then:
```
class DefaultWorkloadApiClient(WorkloadApiClient):
    def __init__(
        self,
        spiffe_socket: str = None,
        grpc_client: GrpcClient = cast(GrpcClient, DefaultGrpcClient),
    ) -> None:
...
    self. grpc_client = grpc_client
...
```

That way we can achieve the goal of using a mock of the GrpcClient for unit testing.

I might be missing something, if that is the case, sorry if this comment make things confusing.

File: Pipfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (scotte) [reply Pipfile:8]: Those lines were already there, all I did was bump the version. I do not know if they need to be in both sections...

File: src/pyspiffe/workloadapi/default_workload_api_client.py (hunk 4 above)
Comment (scotte) [reply src/pyspiffe/workloadapi/default_workload_api_client.py:50]: Maybe? That's why I published this PR. :-) I'm not sure the best approach myself.

But, at this point I want to see what @maxlambrecht has come up with, which will change all of this anyway, I think.

File: Pipfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (scotte) [reply Pipfile:8]: I looked this up - indeed everything in `[packages]` is used for *both* packaging and dev, so these can be removed from `[dev-packages]`. Thanks!

[the remark this answers is not in the document]
Comment (scotte) [reply Pipfile:9]: Fixed, and I sorted dependencies in both sections.

[the remark this answers is not in the document]
Comment (Andres-GC) [reply Pipfile:9]: Great. Thanks Scott.

**Comment by scotte** [comment]
Closing, superseded by @maxlambrecht work in #56

**Pull request closed without merging**